walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,80 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Create arrays to hold all the times attendees have meetings and the free times of all attendees
+    ArrayList<TimeRange> freeTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+
+    // Add each event's time range to busyTimes list if one of the attendees in the request is in that event
+    for (Event event: events) {
+      Set<String> eventAttendees = new HashSet<String>(event.getAttendees());
+      Set<String> requestAttendees = new HashSet<String>(request.getAttendees());
+      if (eventAttendees.removeAll(requestAttendees)) {
+        busyTimes.add(event.getWhen());
+      }
+    }
+
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+
+    // If there are no times when attendees have meetings, check if the duration is less than 24 hrs and
+    // if so then return a time range encompassing the entire day
+    if (busyTimes.size() == 0) {
+      TimeRange wholeDay = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true);
+      if (wholeDay.duration() >= request.getDuration()) {
+        freeTimes.add(wholeDay);
+      }
+      return freeTimes;
+    }
+
+    // Need to filter the busyTimes list so that all overlaps are removed and the list is modified accordingly
+    int i = 0;
+    while (i < busyTimes.size() - 1) {
+      if (busyTimes.get(i).overlaps(busyTimes.get(i+1))) {
+        if (!(busyTimes.get(i).contains(busyTimes.get(i+1)))) {
+          busyTimes.set(i, TimeRange.fromStartEnd(busyTimes.get(i).start(), busyTimes.get(i+1).end(), false));
```

Comment:
Shouldn't this be busyTimes.set(i, TimeRange.fromStartEnd(busyTimes.get(i).start(), busyTimes.get(i+1).start(), false));?

And looks like you could do i++ after this.

Reply:
The reason it is the end of the second is because I want to replace any overlaps with a single TimeRange that encompasses both the TimeRanges. So if we have [0800, 1000] and [0930, 1100] then I want to replace them with [0800, 1100]